### PR TITLE
Adding an option to force EC2 instance ID as hostname

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?
In some cases, users might want to force EC2 instance-id as an hostname.
This is needed when autoscaling a VM with a non standard EC2 hostname
(ie: not starting with "ip-", "domu", "ec2amaz-").

Due to legacy logic GCP and Azure already have the priority when
choosing an hostname but not AWS. On EC2 the OS hostname would be
picked. We need to be backward compatible for the hostname detection so
'ec2_prioritize_instance_id_as_hostname' is introduced.

### Describe how to test/QA your changes

1. Spin a EC2 instance and change the hostname to something else than the default one (ie: something not starting with "ip-", "domu", "ec2amaz-").
2. Run the Agent and check that the hostname is pulled from the OS (the one you set in step `1.`
3. Now set `ec2_prioritize_instance_id_as_hostname` to true and check that the hostname is the ec2 instance ID

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -346,6 +346,9 @@ func InitConfig(config Config) {
 	// compatibility with Agent5 behavior/win
 	config.BindEnvAndSetDefault("hostname_fqdn", false)
 
+	// This is used to bypass the hostname detection logic and force the EC2 instance ID as a hostname.
+	config.BindEnvAndSetDefault("hostname_force_ec2_instance_id", false)
```
Comment (sgnn7) [inline pkg/config/config.go:350]: I think this may be better grouped/prefixed with the [other `ec2_` vars](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config.go#L680-L683). WDYT?

File: pkg/util/hostname.go [hunk 2]
```diff
@@ -159,6 +159,25 @@ func GetHostnameData(ctx context.Context) (HostnameData, error) {
...
+			return saveHostnameData(cacheHostnameKey, hostName, "aws"), nil
+		}
+
+		expErr := new(expvar.String)
+		expErr.Set(err.Error())
+		hostnameErrors.Set("forced EC2 hostname", expErr)
+		// We want to prevent the agent from starting with another hostname if 'hostname_force_ec2_instance_id'
+		// is enabled.
+		return HostnameData{}, fmt.Errorf("'hostname_force_ec2_instance_id' is enable and EC2 instanceID could not be fetched: %s", err)
+	}
```
Comment (sgnn7) [inline pkg/util/hostname.go:179]: I think this block should be combined with the [EC2 hostname eval](https://github.com/DataDog/datadog-agent/blob/main/pkg/util/hostname.go#L272-L288) since both blocks are pretty much doing the same thing.

**Review by sgnn7 — COMMENTED** [review]
@hush-hush Just a couple of comments

File: pkg/config/config.go (hunk 1 above)
Comment (hush-hush) [reply pkg/config/config.go:350]: I actually think it's more related to the hostname detection that how we use EC2.

File: pkg/util/hostname.go (hunk 2 above)
Comment (djmitche) [reply pkg/util/hostname.go:179]: We _really_ need a matrix test for this :(

Comment (sgnn7) [reply pkg/util/hostname.go:179]: > that's the issue of this whole function: lot of tricky edge case. We actually want to force EC2 so we want to check is first thing before we use other providers like the configuration.

I think the fall-through for the new logic should still be the same just that there will be another case tacked on to `if ecs.IsECSInstance() || ec2.IsDefaultHostname(hostName)` case. I believe (if I'm reading the flow correctly) that even if there's a FQDN and OS hostname, we don't early return until we go through the EC2 logic which will should override the hostname as expected.

I guess my overall point is that we use this fall-through for the same logic to use instance IDs instead of hostnames already (when it's a default EC2 hostname) so I wouldn't expect that we would be introducing any gotcha/unexpected new behaviors with the logic change occurring in the EC2 section.

File: pkg/config/config.go (hunk 1 above)
Comment (hush-hush) [reply pkg/config/config.go:350]: Good points :+1:

File: pkg/util/hostname.go (hunk 2 above)
Comment (hush-hush) [reply pkg/util/hostname.go:179]: This would mean that we don't enforce EC2 instance ID but only prioritize it above OS/FQDN. I think the config name is misleading in that case. WDYT ?

Comment (hush-hush) [reply pkg/util/hostname.go]: After our discussion I merged it with the current EC2 logic and updated the configuration name

**Review by sgnn7 — APPROVED** [review]
🚀

**Comment by sgnn7** [comment]
@hush-hush Just a heads up that the unit test is failing in CI afaict

**Comment by olivielpeau** [comment]
@hush-hush I feel like this option should be documented in the config template (and mentioned as a changelog entry), any reason not to?
